// @hubspot/app-functions-dev-server/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildServiceConfiguration = exports.loadFunctionInfo = exports.loadMultiFunctionInfo = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const constants_1 = require("./constants");
// @ts-expect-error type declaration not available
const http_1 = __importDefault(require("@hubspot/cli-lib/http"));
// @ts-expect-error type declaration not available
const cli_lib_1 = require("@hubspot/cli-lib");
const errors_1 = require("./errors");
function loadMultiFunctionInfo(config) {
    const srcDir = path_1.default.join(config.app.path, constants_1.APP_FUNCTIONS_DIRNAME);
    if (!fs_1.default.existsSync(srcDir)) {
        throw new Error(`Cannot locate app function directory: ${srcDir}`);
    }
    const stats = fs_1.default.lstatSync(srcDir);
    if (!stats.isDirectory()) {
        throw new Error(`Invalid app function directory: ${srcDir}`);
    }
    const manifestPath = path_1.default.join(srcDir, constants_1.APP_FUNCTIONS_MANIFEST_FILENAME);
    const manifest = JSON.parse(fs_1.default.readFileSync(manifestPath, {
        encoding: 'utf-8',
    }));
    const { appFunctions = {} } = manifest;
    // TODO validation
    return { appFunctions, srcDir };
}
exports.loadMultiFunctionInfo = loadMultiFunctionInfo;
function loadFunctionInfo(config, functionName, appId) {
    const { logger } = config;
    const { appFunctions, srcDir } = loadMultiFunctionInfo(config);
    const metadata = appFunctions[functionName];
    if (!metadata) {
        const manifestPath = path_1.default.join(srcDir, constants_1.APP_FUNCTIONS_MANIFEST_FILENAME);
        logger.error(`App function "${functionName}" is not declared in ${manifestPath}`);
        throw new errors_1.ExecutionError({
            reason: errors_1.Reason.FunctionNotFound,
            functionName,
            appId,
        });
    }
    // TODO validation
    return Object.assign(Object.assign({}, metadata), { srcDir, name: functionName });
}
exports.loadFunctionInfo = loadFunctionInfo;
function buildServiceConfiguration(appPath, options) {
    var _a, _b, _c, _d;
    return {
        accountId: (_a = options.accountId) !== null && _a !== void 0 ? _a : (0, cli_lib_1.getAccountId)(),
        app: { path: appPath },
        functionTimeoutMs: (_b = options.functionTimeoutMs) !== null && _b !== void 0 ? _b : constants_1.EXECUTION_TIMEOUT_MS,
        httpClient: (_c = options.httpClient) !== null && _c !== void 0 ? _c : http_1.default,
        logger: (_d = options.logger) !== null && _d !== void 0 ? _d : console,
    };
}
exports.buildServiceConfiguration = buildServiceConfiguration;
